### PR TITLE
Feat: Registration Discount codes

### PR DESCRIPTION
**Pull request by kirkas**
Introducing Discount Codes for Basenames.

This PR is addresses two body of works.

### 1. Promo Code

**Get the code from the url `?code=...`**

See `searchParams?.code`  in `apps/web/app/(basenames)/names/page.tsx`. 

**Pass the code to the attestations hooks**

Introducing `useDiscountCodeAttestations`, fetch data from a new endpoint: `/api/proofs/discountCode`. This endpoints compare, validate & sign the code.

From here the existing attestations logic handles the discount automatically.

### 2. RegisterName refactor

We've had issues with the register flow for Smart Wallet, which led to a small refactor. 

This flow is the first one we built, and we've since introduced helpful hooks like `useWriteContractsWithLogs` and `useWriteContractWithReceipt` for better, safer tracking of web3 transactions.

**Changes to `RegistrationContext`**

To avoid rendering race condition and other state / logic mishaps, we moved all the shared logic and state to the Context. This includes `useRegisterNameCallback`, `setYears`, `price`, etc..

**Changes to `RegistrationForm`**

Since the context holds the logic and state, this greatly simplify this components. 

**Changes to `useRegisterNameCallback`**

This hook holds all the logic regarding initiating & tracking transactions, including error and analytics. This simplifies and consolidate Basename registration contract logic into one place.

**Changes to `RegistrationProfileForm`**

Fix which name you are editing.

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

**Review by ricardoMogg — DISMISSED** [review]

File: apps/web/pages/api/proofs/discountCode/consume/index.ts
Comment (JFrankfurt) [inline apps/web/pages/api/proofs/discountCode/consume/index.ts:26]:
```suggestion
```

File: apps/web/pages/api/proofs/discountCode/consume/index.ts
Comment (JFrankfurt) [inline apps/web/pages/api/proofs/discountCode/consume/index.ts:7]: can I just block this endpoint and redeem the code repeatedly?

File: apps/web/src/components/Basenames/RegistrationContext.tsx
Comment (JFrankfurt) [inline apps/web/src/components/Basenames/RegistrationContext.tsx:237]: double checking you don't want this after request completion

**Comment by JFrankfurt** [comment]
lgtm but we do need that prod address uncommented before merge right?

**Review by zencephalon — DISMISSED** [review]

**Comment by kirkas** [comment]
@JFrankfurt yes waiting for contract on mainnet before mergin this. I'll also do a sepolia/codeflow test deploy first

**Review by JFrankfurt — APPROVED** [review]

**Comment by cb-heimdall** [comment]
Review Error for JFrankfurt @ 2024-10-18 20:26:44 UTC
User failed mfa authentication, see [go/mfa-help](http://go/mfa-help)

**Review by JFrankfurt — APPROVED** [review]

**Pull request merged**
